zkvm/src/ops.rs

```diff
@@ -399,4 +400,55 @@ impl Program {
         self.0.push(Instruction::Cloak(m, n));
         self
     }
+
+    /// Takes predicate and closure to add choose operations for
+    /// predicate tree traversal.
+    pub fn choose_predicate<F, T>(
+        &mut self,
+        pred: Predicate,
+        choose_fn: F,
+    ) -> Result<&mut Program, VMError>
+    where
+        F: FnOnce(PredicateTree) -> Result<T, VMError>,
+    {
+        choose_fn(PredicateTree {
+            prog: self,
+            pred: pred,
+        })?;
+        Ok(self)
+    }
+}
+
+/// Adds data and instructions to traverse a predicate tree.
+pub struct PredicateTree<'a> {
+    prog: &'a mut Program,
+    pred: Predicate,
+}
+
+impl<'a> PredicateTree<'a> {
+    /// Left Predicate branch
+    pub fn left(self) -> Result<Self, VMError> {
+        let (l, r) = self.pred.to_disjunction()?;
+        let prog = self.prog;
+        prog.push(l.clone()).push(r).left();
+
+        Ok(Self { pred: l, prog })
+    }
+
+    /// Right Predicate branch
+    pub fn right(self) -> Result<Self, VMError> {
+        let (l, r) = self.pred.to_disjunction()?;
+        let prog = self.prog;
+        prog.push(l).push(r.clone()).right();
+
+        Ok(Self { pred: r, prog })
+    }
+
+    /// Pushes program to the stack and calls the contract protected
+    /// by the program predicate.
+    pub fn call(self) -> Result<&'a mut Program, VMError> {
```

Comment:
```suggestion
    pub fn call(self) -> Result<(), VMError> {
```

Reply:
the program is "owned" by the user of PredicateTree and we don't need to mix concerns in the chain-calling: we chain-call on PredicateTree only, and terminate the chain here to avoid any confusion.